In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score, recall_score, roc_auc_score
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.model_selection import StratifiedKFold

import pandas as pd

# Specify the file path
file_path = "churn.csv"

# Read the CSV file into a DataFrame
data = pd.read_csv(file_path)

In [2]:
data

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0     Female              0     Yes         No       1           No   
1       Male              0      No         No      34          Yes   
2       Male              0      No         No       2          Yes   
3       Male              0      No         No      45           No   
4     Female              0      No         No       2          Yes   
...      ...            ...     ...        ...     ...          ...   
7027    Male              0     Yes        Yes      24          Yes   
7028  Female              0     Yes        Yes      72          Yes   
7029  Female              0     Yes        Yes      11           No   
7030    Male              1     Yes         No       4          Yes   
7031    Male              0      No         No      66          Yes   

         MultipleLines InternetService OnlineSecurity OnlineBackup  \
0     No phone service             DSL             No          Yes   
1                   No             DSL            Yes           No   
2                   No             DSL            Yes          Yes   
3     No phone service             DSL            Yes           No   
4                   No     Fiber optic             No           No   
...                ...             ...            ...          ...   
7027               Yes             DSL            Yes           No   
7028               Yes     Fiber optic             No          Yes   
7029  No phone service             DSL            Yes           No   
7030               Yes     Fiber optic             No           No   
7031                No     Fiber optic            Yes           No   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0                  No          No          No              No  Month-to-month   
1                 Yes          No          No              No        One year   
2                  No          No          No              No  Month-to-month   
3                 Yes         Yes          No              No        One year   
4                  No          No          No              No  Month-to-month   
...               ...         ...         ...             ...             ...   
7027              Yes         Yes         Yes             Yes        One year   
7028              Yes          No         Yes             Yes        One year   
7029               No          No          No              No  Month-to-month   
7030               No          No          No              No  Month-to-month   
7031              Yes         Yes         Yes             Yes        Two year   

     PaperlessBilling              PaymentMethod  MonthlyCharges Churn  
0                 Yes           Electronic check           29.85    No  
1                  No               Mailed check           56.95    No  
2                 Yes               Mailed check           53.85   Yes  
3                  No  Bank transfer (automatic)           42.30    No  
4                 Yes           Electronic check           70.70   Yes  
...               ...                        ...             ...   ...  
7027              Yes               Mailed check           84.80    No  
7028              Yes    Credit card (automatic)          103.20    No  
7029              Yes           Electronic check           29.60    No  
7030              Yes               Mailed check           74.40   Yes  
7031              Yes  Bank transfer (automatic)          105.65    No  

[7032 rows x 19 columns]

## Question 1

In [3]:
# Encode categorical variables
label_encoders = {}
for column in data.select_dtypes(include=['object']).columns:
    if column != 'Churn':
        le = LabelEncoder()
        data[column] = le.fit_transform(data[column])
        label_encoders[column] = le

# Encode target variable
data['Churn'] = data['Churn'].apply(lambda x: 1 if x == 'Yes' else 0)

# Scale data
scaler = StandardScaler()
data[data.columns[:-1]] = scaler.fit_transform(data[data.columns[:-1]])

# Split features and target
X = data.drop('Churn', axis=1)
y = data['Churn']

# Initialize the Logistic Regression Model
log_reg = LogisticRegression()

# Perform 10-fold cross-validation
cv = StratifiedKFold(n_splits=10)
acc_scores = cross_val_score(log_reg, X, y, cv=cv, scoring='accuracy')
precision_scores = cross_val_score(log_reg, X, y, cv=cv, scoring='precision')
recall_scores = cross_val_score(log_reg, X, y, cv=cv, scoring='recall')
auc_scores = cross_val_score(log_reg, X, y, cv=cv, scoring='roc_auc')


from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import StratifiedKFold, cross_val_score

# Encode categorical variables
label_encoders = {}
for column in data.columns:
    if data[column].dtype == 'object' and column != 'Churn':  # Exclude 'Churn' as it's already processed
        le = LabelEncoder()
        data[column] = le.fit_transform(data[column])
        label_encoders[column] = le

# Scale numeric data
scaler = StandardScaler()
numerical_columns = data.select_dtypes(include=['float64', 'int64']).columns.tolist()
data[numerical_columns] = scaler.fit_transform(data[numerical_columns])

# Split features and target
X = data.drop('Churn', axis=1)
y = data['Churn'].astype(int)  # Ensure Churn is treated as an integer

# Initialize the Logistic Regression Model
log_reg1 = LogisticRegression()

# Perform 10-fold cross-validation
cv = StratifiedKFold(n_splits=10)
acc_scores = cross_val_score(log_reg1, X, y, cv=cv, scoring='accuracy')
precision_scores = cross_val_score(log_reg1, X, y, cv=cv, scoring='precision')
recall_scores = cross_val_score(log_reg1, X, y, cv=cv, scoring='recall')
auc_scores = cross_val_score(log_reg1, X, y, cv=cv, scoring='roc_auc')

# Output the evaluation results
print(f'Accuracy: {acc_scores.mean()}')
print(f'Precision: {precision_scores.mean()}')
print(f'Recall: {recall_scores.mean()}')
print(f'AUC: {auc_scores.mean()}')


Accuracy: 0.8024747833958361
Precision: 0.6543130529231973
Recall: 0.543056753493186
AUC: 0.8418610498322716


*Accuracy (80.25%) Interpretation: The model correctly predicts whether a customer will churn or not 80.25% of the time. It provides a general idea of overall performance but needs to be viewed alongside other metrics for a complete picture. 

*Precision (65.43%) Interpretation: When the model predicts that a customer will churn, it is correct 65.43% of the time. This indicates a moderate level of reliability in the positive predictions (churn predictions). 

*Recall (54.31%) Interpretation: The model identifies 54.31% of all actual churners. This suggests it misses a significant number of churners, highlighting a potential area for improvement. 

*AUC (84.19%) Interpretation: The high AUC indicates that the model is quite good at distinguishing between customers who will churn and those who will not. This suggests effective class separation capabilities.

## Question 2

From Task 1

In [4]:
import pandas as pd
from sklearn.model_selection import train_test_split, StratifiedKFold, cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.impute import SimpleImputer

# Load data into a DataFrame named 'df'
df = pd.read_csv('attrition.csv')

# Encode 'Attrition' assuming it contains string labels 'Yes' and 'No'
df['Attrition'] = df['Attrition'].apply(lambda x: 1 if x == 'Yes' else 0)

# Encode all categorical variables
categorical_cols = df.select_dtypes(include=['object']).columns
for column in categorical_cols:
    df[column] = LabelEncoder().fit_transform(df[column])

# Handle missing values
imputer = SimpleImputer(strategy='most_frequent')
df = pd.DataFrame(imputer.fit_transform(df), columns=df.columns)

# Scaling the data
scaler = StandardScaler()
df[df.columns.difference(['Attrition'])] = scaler.fit_transform(df[df.columns.difference(['Attrition'])])

# Define features and target
X = df.drop('Attrition', axis=1)
y = df['Attrition'].astype(int)  # Ensure 'Attrition' is integer for modeling

# Initialize and train Logistic Regression with increased max_iter and a different solver if needed
log_reg = LogisticRegression(max_iter=1000, solver='lbfgs')  # Increase max_iter, keep lbfgs or try 'saga' if still not converging
log_reg.fit(X, y)  # Fit the model

# Evaluate the model using cross-validation
cv = StratifiedKFold(n_splits=10)
scores = cross_val_score(log_reg, X, y, cv=cv, scoring='roc_auc')
print(f"\nModel AUC: {scores.mean():.4f}")



Model AUC: 0.8197


The Model AUC (Area Under the Curve) of 0.8197 is quite strong, indicating that your logistic regression model is performing well in distinguishing between employees who are likely to leave and those who are likely to stay. An AUC score close to 0.82 means that there is a high chance that the model will be able to differentiate between positive (attrition) and negative (no attrition) classes effectively.

In [5]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.impute import SimpleImputer

# Load the churn dataset
df_churn = pd.read_csv('churn.csv')

# Explicitly identify categorical and numerical columns if known
categorical_cols_churn = ['gender', 'Partner', 'Dependents', 'PhoneService', 'MultipleLines',
                          'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection',
                          'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
                          'PaymentMethod']  # Example, adjust according to your dataset

# Assuming 'MonthlyCharges' and 'tenure' are numerical that need scaling
numerical_cols_churn = ['MonthlyCharges', 'tenure']

# Encode categorical variables
for column in categorical_cols_churn:
    if column in df_churn.columns:
        df_churn[column] = LabelEncoder().fit_transform(df_churn[column])

# Impute missing values using a strategy appropriate for each column type
imputer_cat = SimpleImputer(strategy='most_frequent')
df_churn[categorical_cols_churn] = imputer_cat.fit_transform(df_churn[categorical_cols_churn])

imputer_num = SimpleImputer(strategy='mean')
df_churn[numerical_cols_churn] = imputer_num.fit_transform(df_churn[numerical_cols_churn])

# Scale numerical data
scaler = StandardScaler()
df_churn[numerical_cols_churn] = scaler.fit_transform(df_churn[numerical_cols_churn])

In [6]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.impute import SimpleImputer

# Let's assume 'df' is the DataFrame that was used to train the model in Task 1
# Load data into a DataFrame named 'df'
df = pd.read_csv('attrition.csv')

# Encode 'Attrition' assuming it contains string labels 'Yes' and 'No'
df['Attrition'] = df['Attrition'].apply(lambda x: 1 if x == 'Yes' else 0)

# Encode all categorical variables
categorical_cols = df.select_dtypes(include=['object']).columns
for column in categorical_cols:
    df[column] = LabelEncoder().fit_transform(df[column])

# Handle missing values
imputer = SimpleImputer(strategy='most_frequent')
df = pd.DataFrame(imputer.fit_transform(df), columns=df.columns)

# Scaling the data
scaler = StandardScaler()
df[df.columns.difference(['Attrition'])] = scaler.fit_transform(df[df.columns.difference(['Attrition'])])

# Define features and target
X = df.drop('Attrition', axis=1)
y = df['Attrition'].astype(int)  # Ensure 'Attrition' is integer for modeling

# Assuming 'log_reg' is the trained Logistic Regression model
# If it's not defined, here's how you might have trained it:
log_reg = LogisticRegression(max_iter=1000)
log_reg.fit(X, y)

# Predict using the trained model
predicted_classes = log_reg.predict(X)
predicted_probabilities = log_reg.predict_proba(X)[:, 1]

In [7]:
# Assuming `X` is your features matrix from the dataset used to fit the model
predicted_classes = log_reg.predict(X)
predicted_probabilities = log_reg.predict_proba(X)[:, 1]

# Print all predictions (class labels and probabilities)
for idx, (cls, prob) in enumerate(zip(predicted_classes, predicted_probabilities)):
    print(f"Employee {idx + 1}: Predicted Class - {cls}, Predicted Probability - {prob:.4f}")


Employee 1: Predicted Class - 1, Predicted Probability - 0.7087
Employee 2: Predicted Class - 0, Predicted Probability - 0.0161
Employee 3: Predicted Class - 1, Predicted Probability - 0.5785
Employee 4: Predicted Class - 0, Predicted Probability - 0.1588
Employee 5: Predicted Class - 0, Predicted Probability - 0.4301
Employee 6: Predicted Class - 0, Predicted Probability - 0.0314
Employee 7: Predicted Class - 0, Predicted Probability - 0.1996
Employee 8: Predicted Class - 0, Predicted Probability - 0.0630
Employee 9: Predicted Class - 0, Predicted Probability - 0.0275
Employee 10: Predicted Class - 0, Predicted Probability - 0.1223
Employee 11: Predicted Class - 0, Predicted Probability - 0.0722
Employee 12: Predicted Class - 0, Predicted Probability - 0.1681
Employee 13: Predicted Class - 0, Predicted Probability - 0.2189
Employee 14: Predicted Class - 0, Predicted Probability - 0.0528
Employee 15: Predicted Class - 1, Predicted Probability - 0.8645
Employee 16: Predicted Class - 0, 

## Question 3

In [8]:
# Load the churn dataset
df_churn = pd.read_csv('churn.csv')

In [9]:
# Calculate the Contract Value (CV) for each customer
df_churn['CV'] = df_churn['MonthlyCharges'] * 12

# Total costs when doing nothing
# All churners churn, no offers sent, so no -205 reduction
cost_do_nothing = df_churn[df_churn['Churn'] == 1]['CV'].sum()

# Total costs when sending offers to everyone
# Costs incurred from all not churning receiving unnecessary offers
cost_send_everyone = len(df_churn) * 205

# Print calculated costs
print(f"Cost of Doing Nothing: ${cost_do_nothing:,.2f}")
print(f"Cost of Sending Offers to Everyone: ${cost_send_everyone:,.2f}")


Cost of Doing Nothing: $0.00
Cost of Sending Offers to Everyone: $1,441,560.00


*Cost of Doing Nothing: All churners leave, the company loses the full CV of these customers without subtracting 205 since no retention efforts were made.  In this scenario, the company opts not to engage in any proactive retention efforts. The cost is calculated based on the total contract value (CV) that would be lost due to customers who churn. This represents a scenario where the company takes no action to mitigate churn, resulting in a total financial loss equivalent to the sum of the contract values of all churners.

*Cost of Sending Offers to Everyone: Every customer receives an offer, costing the company 205 per customer. This cost incorporates both the 200 for the offer and 5 for operational costs, straightforwardly applied across the entire customer base.  Here, every customer, regardless of their likelihood to churn, receives an offer. This approach leads to a significant cost as it involves spending 205 per customer (inclusive of the 200 offer cost and 5 operational costs), applied across the entire customer base.

Conclusion:

Both baseline strategies are economically inefficient. The "Doing Nothing" approach fails to protect against revenue loss from churn, whereas "Sending Offers to Everyone" results in excessive and unnecessary expenditures.

Implementing a data-driven strategy using predictive modeling can significantly enhance decision-making. By identifying customers at high risk of churn, the company can selectively target their retention efforts, potentially saving substantial amounts on unwarranted offers and focusing resources where they are most likely to yield a return in customer retention.

## Question 4

In [10]:
# Load the churn dataset
df = pd.read_csv('churn.csv')

In [11]:
# Add a new column 'Customer_ID' with unique IDs for each row
df['Customer_ID'] = range(1, len(data) + 1)

In [12]:
# Calculate contract value (CV) for each customer and add it as a new column
df['CV'] = df['MonthlyCharges'] * 12

In [13]:
df

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0     Female              0     Yes         No       1           No   
1       Male              0      No         No      34          Yes   
2       Male              0      No         No       2          Yes   
3       Male              0      No         No      45           No   
4     Female              0      No         No       2          Yes   
...      ...            ...     ...        ...     ...          ...   
7027    Male              0     Yes        Yes      24          Yes   
7028  Female              0     Yes        Yes      72          Yes   
7029  Female              0     Yes        Yes      11           No   
7030    Male              1     Yes         No       4          Yes   
7031    Male              0      No         No      66          Yes   

         MultipleLines InternetService OnlineSecurity OnlineBackup  ...  \
0     No phone service             DSL             No          Yes  ...   
1                   No             DSL            Yes           No  ...   
2                   No             DSL            Yes          Yes  ...   
3     No phone service             DSL            Yes           No  ...   
4                   No     Fiber optic             No           No  ...   
...                ...             ...            ...          ...  ...   
7027               Yes             DSL            Yes           No  ...   
7028               Yes     Fiber optic             No          Yes  ...   
7029  No phone service             DSL            Yes           No  ...   
7030               Yes     Fiber optic             No           No  ...   
7031                No     Fiber optic            Yes           No  ...   

     TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0             No          No              No  Month-to-month              Yes   
1             No          No              No        One year               No   
2             No          No              No  Month-to-month              Yes   
3            Yes          No              No        One year               No   
4             No          No              No  Month-to-month              Yes   
...          ...         ...             ...             ...              ...   
7027         Yes         Yes             Yes        One year              Yes   
7028          No         Yes             Yes        One year              Yes   
7029          No          No              No  Month-to-month              Yes   
7030          No          No              No  Month-to-month              Yes   
7031         Yes         Yes             Yes        Two year              Yes   

                  PaymentMethod MonthlyCharges  Churn Customer_ID      CV  
0              Electronic check          29.85     No           1   358.2  
1                  Mailed check          56.95     No           2   683.4  
2                  Mailed check          53.85    Yes           3   646.2  
3     Bank transfer (automatic)          42.30     No           4   507.6  
4              Electronic check          70.70    Yes           5   848.4  
...                         ...            ...    ...         ...     ...  
7027               Mailed check          84.80     No        7028  1017.6  
7028    Credit card (automatic)         103.20     No        7029  1238.4  
7029           Electronic check          29.60     No        7030   355.2  
7030               Mailed check          74.40    Yes        7031   892.8  
7031  Bank transfer (automatic)         105.65     No        7032  1267.8  

[7032 rows x 21 columns]

In [14]:
# Fit the logistic regression model
log_reg1.fit(X, y)

# Predict probabilities of churn for each customer using the logistic regression model
prob_churn = log_reg1.predict_proba(X)
# The predicted probabilities will be a 2D array, where each row represents a sample and each column represents a class.
# We are interested in the probability of class 1, which corresponds to churn.
prob_churn_class_1 = prob_churn[:, 1]

# Print the probabilities of churn for each row
print("Probabilities of churn for each row:")
print(prob_churn_class_1)

Probabilities of churn for each row:
[0.70865354 0.0160979  0.57850402 ... 0.34016576 0.01494566 0.03949416]


# Question 5

We can develop a more sophisticated and effective retention strategy that takes into account the likelihood of customer acceptance and improves customer retention rates while minimizing churn.

Experimentation and A/B Testing: Conduct controlled experiments to test different retention strategies and measure their impact on customer churn. By randomly assigning customers to different treatment groups (e.g., receiving different retention offers), we can evaluate the effectiveness of each strategy and identify the most successful approaches.